common/Renderers/DevHomeActionSet.cs

```diff
@@ -0,0 +1,137 @@
+/// within its own UI the to top level action set in the adaptive card.
+/// </summary>
+/// <remarks>
+/// It is expected that the adaptive card will have a top level action set with buttons that Dev Home can link to.
+/// </remarks>
+public class DevHomeActionSet : IDevHomeActionSetRender
+{
+    /// <summary>
+    /// Gets the visibility of the action in Dev Home's UI.
+    /// </summary>
+    private readonly TopLevelCardActionSetVisibility _actionSetVisibility;
+
+    /// <summary>
+    /// Gets the adaptive card object that will invoke an action within the action set.
+    /// </summary>
+    public AdaptiveActionInvoker? ActionButtonInvoker { get; private set; }
+
+    public AdaptiveCard? OriginalAdaptiveCard { get; private set; }
+
+    private Dictionary<string, IAdaptiveActionElement> ActionButtonMap { get; } = new();
+
+    public DevHomeActionSet(TopLevelCardActionSetVisibility cardActionSetVisibility)
+    {
+        _actionSetVisibility = cardActionSetVisibility;
+    }
+
+    public UIElement? Render(IAdaptiveCardElement element, AdaptiveRenderContext context, AdaptiveRenderArgs renderArgs)
+    {
+        ActionButtonMap.Clear();
+
+        if (element is AdaptiveActionSet actionSet)
+        {
+            foreach (var action in actionSet.Actions)
+            {
+                if (action is AdaptiveExecuteAction executeAction)
+                {
+                    context.LinkSubmitActionToCard(executeAction, renderArgs);
+                }
+                else if (action is AdaptiveSubmitAction submitAction)
+                {
+                    context.LinkSubmitActionToCard(submitAction, renderArgs);
+                }
+
+                ActionButtonMap.TryAdd(action.Id, action);
+            }
+
+            ActionButtonInvoker = context.ActionInvoker;
+            OriginalAdaptiveCard = renderArgs.ParentCard;
+
+            if (_actionSetVisibility == TopLevelCardActionSetVisibility.Hidden)
+            {
+                // the page in Dev Home does not want to show the action set in the card.
+                // So we return null to prevent the adaptive card buttons from appearing.
+                // Invoking the button from Dev Home will then trigger the action in the adaptive card.
+                return null;
+            }
+        }
+
+        var renderer = new AdaptiveActionSetRenderer();
+        return renderer.Render(element, context, renderArgs);
+    }
+
+    /// <summary>
+    /// Invokes an adaptive card action from anywhere within Dev Home, like a method in a view Model for example.
+    /// A boolean is returned with the validation result of the card. We still send the action event so the adaptive
+    /// cards UI updates with error information.
+    /// </summary>
+    /// <returns>
+    /// A boolean indicating whether validation for the card passed or failed.
+    /// </returns>
+    public bool TryValidateAndInitiateAction(string buttonId, AdaptiveInputs userInputs)
```

Comment:
https://github.com/microsoft/devhome/pull/2493/files#diff-9aa0e204d624670b424e4a4de48b0aa416de4f630d62b806e408f15274943c18R293 looks like `userInputs` will be `null` if the ViewModel is `null`.

crash if null(?)